worker/upload.go

```diff
@@ -0,0 +1,1154 @@
+package worker
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/montanaflynn/stats"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+	rhpv2 "go.sia.tech/core/rhp/v2"
+	"go.sia.tech/core/types"
+	"go.sia.tech/renterd/api"
+	"go.sia.tech/renterd/object"
+	"go.sia.tech/renterd/tracing"
+	"lukechampine.com/frand"
+)
+
+var (
+	errNoFreeQueue    = errors.New("no free queue")
+	errNotEnoughHosts = errors.New("not enough hosts to support requested redundancy")
+)
+
+type (
+	uploadID [8]byte
+
+	uploader struct {
+		hp hostProvider
+		rl revisionLocker
+
+		overdriveTimeout time.Duration
+		maxOverdrive     uint64
+
+		statsOverdrive *dataPoints
+		statsSpeed     *dataPoints
+
+		stopChan chan struct{}
+
+		mu        sync.Mutex
+		contracts []*uploadQueue
+		completed map[uploadID]map[uploadID]map[types.FileContractID]struct{}
+		excluded  map[uploadID]map[types.FileContractID]struct{}
+		history   map[uploadID][]uploadID
+		used      map[uploadID]map[uploadID]map[types.FileContractID]struct{}
+
+		nextSlabTriggers        map[uploadID]chan struct{}
+		sectorCompletedTriggers map[uploadID]chan struct{}
+	}
+
+	uploadStats struct {
+		overdrivePct   float64
+		queuesHealthy  uint64
+		queuesSpeedAvg float64
+		queuesTotal    uint64
+		topTenHosts    []hostStats
+	}
+
+	hostStats struct {
+		hk       types.PublicKey
+		estimate float64
+	}
+
+	uploadQueue struct {
+		fcid       types.FileContractID
+		hk         types.PublicKey
+		siamuxAddr string
+
+		mu                  sync.Mutex
+		bh                  uint64
+		consecutiveFailures uint64
+		queueChan           chan struct{}
+		queue               []*uploadJob
```

Comment:
To me this is the actual `uploadQueue` (a data structure that holds uploads) while the `uploadQueue` would be more of an `uploader` (thing bound to a host that works itself through the queue) and the `uploader` is kind of an `uploadManager` (since it orchestrates all uploaders).

Reply:
This took forever but I changed it ^^